content/blog/2018-06-07-when-to-use-derived-state.md

```diff
@@ -0,0 +1,331 @@
+
+For the simple example above, we could "fix" the problem of unexpected re-renders using `shouldComponentUpdate`. However in practice, components usually accept multiple props, and our component would re-render if any one of them changed. Even if none of them "changed", props that are functions or objects are often created inline and so will always bypass `shouldComponentUpdate`. For example, what if our component accepted a function to validate the current email address?
+```js
+<EmailInput
+  email={this.props.user.email}
+  validate={this.validateEmail.bind(this)}
+/>
+```
+
+The above example binds the validation callback inline and so it will pass a new function prop every time it renders– effectively bypassing `shouldComponentUpdate` entirely. [Here is a demo](codesandbox://when-to-use-derived-state/derived-state-anti-pattern) that uses a timer to illustrate this problem. Because it might break at any time your component needs to accept new props, this design pattern is inherently fragile.
+
+Hopefully it's clear by now why unconditionally overriding state with props is a bad idea. But what if we were to only update the state when the email prop changes? We'll take a look at that pattern next.
+
+### Anti-pattern: Props override state whenever they change
+
+Building on our example above, we could avoid accidentally erasing state by only updating it when `props.email` changes:
+
+```js
+class EmailInput extends Component {
+  state = {
+    prevPropsEmail: this.props.email,
+    email: this.props.email
+  };
+
+  static getDerivedStateFromProps(props, state) {
+    // Any time props.email changes, update state.
+    if (props.email !== state.prevPropsEmail) {
+      return {
+        prevPropsEmail: props.email,
+        email: props.email
+      };
+    }
+    return null;
+  }
+  
+  // ...
+}
+```
+
+We've just made a big improvement. Now our component will no longer erase what we've typed every time it renders. At this point, what we have is fairly similar to the [scroll direction example](blog/2018/03/27/update-on-async-rendering.html#updating-state-based-on-props) we mentioned before. (It will also more closely mirror how `componentWillReceiveProps` would have been used in the past.)
+
+There is still one subtle problem though, and it's probably easiest to illustrate with an example. Let's say our `EmailInput` component is used inside of an "edit profile" form. The first time the form is rendered, the component will display our email address. Let's say we edited the form (including our email address) but then changed our mind and clicked a "reset" button. At this point, we would expect all form fields to return to their initial values– but the `EmailInput` component **will not be reset**. Do you know why?
+
+The reason for this is that `props.email` never actually changed in the above scenario. Both times the "edit profile" form rendered, it would pass our saved email address via props.
+
```

Comment:
WDYT about adding:

>This problem manifests itself even if you don't have a "Reset" button. For example, imagine a password manager app using the above input implementation. When we navigate between the detail forms for two accounts with the same initial email, the input would fail to reset because the prop value from above is the same! That would be a surprise to the user, as a draft change to one account shouldn't affect another account after navigation.

Reply:
I dig it.